version-negotiation.md

```diff
@@ -0,0 +1,74 @@
+# Explainer: version negotiation for origin policies
+
+This explainer lays out the problem and solution space for how origin policies can be updated, in a privacy-preserving way.
+
+## Use cases and requirements
+
+### Async update ([#10](https://github.com/WICG/origin-policy/issues/10))
+
+Web applications want to be able to decide that a previously-downloaded origin policy, or even no policy at all, is "good enough", and that they don't need to block the page load waiting for the newest origin policy version.
+
+### Sync update
+
+Web applications want to be able to decide that a certain policy is mandatory, and the page cannot load until it arrives and is applied.
+
+Similarly, web applications may have previously sent a bad policy, which they need to replace before any further page loads continue.
+
+### Kill switch
+
+There needs to be a way for a page to remove any previous origin policies, before the page loads. This is essentially a special case of sync update, which doesn't require loading a new policy since the browser already knows what the null policy looks like.
+
+### Privacy-preserving
+
+Any scenario in which the browser communicates previously-sent information back to the server or website can be (ab)used as a cookie. Previous incarnations of origin policy communicated the currently-downloaded policy version. But even the behaviors controlled by origin policy could be used in this manner.
+
+## Proposed model
+
+Conceptually, origin policies are stored in the HTTP cache, under the URL `$origin/.well-known/origin`. In particular, this means they are double-keyed [in the same way as the HTTP cache](https://github.com/WICG/origin-policy/issues/23): that is, by (origin of document, origin of top-document). This prevents them from being used for cross-site tracking.
+
+(Question: should clearing the HTTP cache also clear any cached origin policy? Probably yes. Or should it be tied to clearing cookies?)
```

Comment:
Very much yes.

(At least, on the Chromium side, clearing browser data is considered a privacy tool, and there shouldn't remain a record of visited sites after the cache is cleared. I'm not sure, but I think other browsers see this similarly. The clear data dialog does allow fairly fine-grained controls, but it seems many users are not aware of every detail of browser data storage, and would probably expect something like origin policy to be a part of the cache.)

Reply:
Makes sense. Do you think clearing cookies should have any related effect? (Maybe this is a more general issue where, since the cache can be used as a cookie via timing attacks, clearing the two should be related.)

Reply:
I think the Privacy/UX folks will insist on figuring this out on their won.

My take is that a reasonably advanced user understands what cookies are, and understands what a cache is, and I think those things are some of the things you can choose to clear separately. Since I find it unlikely that many users will understand what an origin policy is, the question becomes: Is origin policy like any of those other, known things, and in that case, clear it whenever that other thing is cleared.

My intuition is that it's mostly like a cache entry. (One could make the argument that it's more like a cookie. Not sure.)

That said, I guess most people will do a full delete anyhow.

Reply:
In Firefox cookies/site data and cache are currently separate, but once we have cache separation per origin I suspect we'll start integrating those from a UI perspective. Users shouldn't have to understand the distinction there.

Reply:
In Firefox cookies/site data and cache are currently separate, but once we have cache separation per origin I suspect we'll start integrating those from a UI perspective. Users shouldn't have to understand the distinction there.